File: app/src/main/java/id/kuato/woahelper/widgets/MountWidget.java
package id.kuato.woahelper.widgets;

import android.annotation.SuppressLint;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.widget.RemoteViews;

import java.util.List;

import id.kuato.woahelper.R;
import id.kuato.woahelper.main.MainActivity;

public class MountWidget extends AppWidgetProvider {
    @SuppressLint("StringFormatInvalid")
    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            Intent intent = new Intent(context, MountWidget.class);
            intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
            intent.setAction(WidgetActivity.ACTION_CLICK);
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_layout);

            PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);

            views.setOnClickPendingIntent(R.id.root, pendingIntent);
            views.setImageViewResource(R.id.image, R.drawable.ic_mnt);

            views.setTextViewText(R.id.text, (MainActivity.isMounted() ? context.getString(R.string.mnt_title, context.getString(R.string.unmountt)) : context.getString(R.string.mnt_title, context.getString(R.string.mountt))));
            appWidgetManager.updateAppWidget(appWidgetId, views);
        }
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
        if (WidgetActivity.ACTION_CLICK.equals(intent.getAction())) {
            int appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_layout);

            Intent i = new Intent(context, WidgetActivity.class);
            i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            i.putExtra("WIDGET_TYPE", "mount");
            context.startActivity(i);

            appWidgetManager.updateAppWidget(appWidgetId, views);
        }
        if ("UPDATE".equals(intent.getAction())) {
            String text = intent.getStringExtra("text");

            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            int[] appWidgetIds = appWidgetManager.getAppWidgetIds(new ComponentName(context, MountWidget.class));
            onUpdate(context, appWidgetManager, appWidgetIds);

            for (int appWidgetId : appWidgetIds) {
                RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_layout);

                views.setTextViewText(R.id.text, text);

                appWidgetManager.updateAppWidget(appWidgetId, views);
            }
        }
    }

    public static void updateText(Context context,  String text) {
        Intent intent = new Intent(context, MountWidget.class);
        intent.setAction("UPDATE");
        intent.putExtra("text", text);
        context.sendBroadcast(intent);
    }
}

File: app/src/main/kotlin/com/woa/helper/widget/MountWidget.kt
package com.woa.helper.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.widget.RemoteViews
import android.widget.Toast
import com.woa.helper.R
import com.woa.helper.util.MountManager
import com.woa.helper.util.ShellManager
import com.woa.helper.util.ShellResult

class MountWidget : AppWidgetProvider() {

    override fun onReceive(context: Context, intent: Intent) {
        when (intent.action) {
            ACTION_MOUNT_TOGGLE -> runAsync(context) {
                val result = if (MountManager.isMounted()) MountManager.unmount() else MountManager.mount()
                val isMounted = MountManager.isMounted()
                updateAllWidgets(context, isMounted)
                val msg = when (result) {
                    is ShellResult.Success -> context.getString(if (isMounted) R.string.mounted else R.string.unmounted)
                    is ShellResult.Error -> "Operation failed: ${result.message}"
                }
                Handler(Looper.getMainLooper()).post { Toast.makeText(context, msg, Toast.LENGTH_SHORT).show() }
            }
            ACTION_UPDATE_WIDGET, AppWidgetManager.ACTION_APPWIDGET_UPDATE -> runAsync(context) {
                updateAllWidgets(context, runCatching { MountManager.isMounted() }.getOrDefault(false))
            }
            else -> super.onReceive(context, intent)
        }
    }

    private fun runAsync(context: Context, block: () -> Unit) {
        val pendingResult = goAsync()
        Thread {
            try {
                ShellManager.init(context.filesDir)
                MountManager.init(context.filesDir, context)
                block()
            } catch (e: Exception) {
                Handler(Looper.getMainLooper()).post {
                    Toast.makeText(context, "Error: ${e.message}", Toast.LENGTH_SHORT).show()
                }
            } finally {
                pendingResult.finish()
            }
        }.start()
    }

    private fun updateAllWidgets(context: Context, isMounted: Boolean) {
        val appWidgetManager = AppWidgetManager.getInstance(context)
        val ids = appWidgetManager.getAppWidgetIds(ComponentName(context, MountWidget::class.java))
        val action = context.getString(if (isMounted) R.string.unmountt else R.string.mountt)
        val views = RemoteViews(context.packageName, R.layout.widget_mount).apply {
            setTextViewText(R.id.widget_text, context.getString(R.string.mnt_title, action))
            setImageViewResource(R.id.widget_icon, R.drawable.mnt)
            setOnClickPendingIntent(R.id.widget_container, clickIntent(context))
        }
        ids.forEach { appWidgetManager.updateAppWidget(it, views) }
    }

    private fun clickIntent(context: Context) = PendingIntent.getBroadcast(
        context, 0, Intent(context, MountWidget::class.java).apply { action = ACTION_MOUNT_TOGGLE },
        PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
    )

    companion object {
        const val ACTION_MOUNT_TOGGLE = "com.woa.helper.widget.ACTION_MOUNT_TOGGLE"
        const val ACTION_UPDATE_WIDGET = "com.woa.helper.widget.ACTION_UPDATE_WIDGET"

        fun requestUpdate(context: Context) {
            context.sendBroadcast(Intent(context, MountWidget::class.java).apply { action = ACTION_UPDATE_WIDGET })
        }
    }
}
